File: apps/cove-workers/Sources/cove_workers/Features/Telemetry/Telemetry.swift
import Foundation

#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

/// Hand-rolled OTLP HTTP JSON exporter (B3-clean — no vendor SDK).
final class Telemetry {
  private(set) var exporterStatus: String = "healthy"
  private let base: String
  private let service: String

  init(ctx: RuntimeContext) {
    var b = ctx.exporterEndpoint
    while b.hasSuffix("/") { b.removeLast() }
    self.base = b
    self.service = ctx.service
  }

  func spanAndLog(name: String) {
    let now = UInt64(Date().timeIntervalSince1970 * 1_000_000_000)
    let traceId = String(format: "%032llx", now % 1_000_000_000_000_000)
    let spanId = String(format: "%016llx", now % 1_000_000_000_000)
    let body =
      "{\"resourceSpans\":[{\"resource\":{\"attributes\":[{\"key\":\"service.name\",\"value\":{\"stringValue\":\"\(service)\"}}]},\"scopeSpans\":[{\"spans\":[{\"traceId\":\"\(traceId)\",\"spanId\":\"\(spanId)\",\"name\":\"\(name)\",\"kind\":1,\"startTimeUnixNano\":\"\(now)\",\"endTimeUnixNano\":\"\(now + 1_000_000)\",\"status\":{\"code\":1}}]}]}]}"
    _ = post(path: "/v1/traces", json: body)
    let log =
      "{\"resourceLogs\":[{\"resource\":{\"attributes\":[{\"key\":\"service.name\",\"value\":{\"stringValue\":\"\(service)\"}}]},\"scopeLogs\":[{\"logRecords\":[{\"timeUnixNano\":\"\(now)\",\"severityNumber\":9,\"body\":{\"stringValue\":\"request\"}}]}]}]}"
    if !post(path: "/v1/logs", json: log) {
      exporterStatus = "degraded"
    }
  }

  func increment(name: String) {
    let now = UInt64(Date().timeIntervalSince1970 * 1_000_000_000)
    let body =
      "{\"resourceMetrics\":[{\"resource\":{\"attributes\":[{\"key\":\"service.name\",\"value\":{\"stringValue\":\"\(service)\"}}]},\"scopeMetrics\":[{\"metrics\":[{\"name\":\"\(name)\",\"sum\":{\"aggregationTemporality\":2,\"isMonotonic\":true,\"dataPoints\":[{\"asInt\":\"1\",\"startTimeUnixNano\":\"\(now)\",\"timeUnixNano\":\"\(now)\"}]}}]}]}]}"
    if !post(path: "/v1/metrics", json: body) {
      exporterStatus = "degraded"
    }
  }

  @discardableResult
  private func post(path: String, json: String) -> Bool {
    guard let url = URL(string: base + path) else { return false }
    var req = URLRequest(url: url)
    req.httpMethod = "POST"
    req.setValue("application/json", forHTTPHeaderField: "Content-Type")
    req.httpBody = json.data(using: .utf8)
    let sem = DispatchSemaphore(value: 0)
    var ok = false
    URLSession.shared.dataTask(with: req) { _, response, error in
      defer { sem.signal() }
      if error != nil { return }
      if let http = response as? HTTPURLResponse {
        ok = (200..<300).contains(http.statusCode)
      }
    }.resume()
    _ = sem.wait(timeout: .now() + 2)
    return ok
  }
}

File: apps/quay-api/src/telemetry.go
package main

import (
	"context"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

type telemetry struct {
	base, service, status string
	client                *http.Client
}

func startOtel(ctx runtimeContext) *telemetry {
	return &telemetry{
		base:    strings.TrimRight(ctx.ExporterEndpoint, "/"),
		service: ctx.Service,
		status:  "healthy",
		client:  &http.Client{Timeout: 2 * time.Second},
	}
}

func (t *telemetry) emit(name string) {
	now := time.Now().UnixNano()
	traceID := fmt.Sprintf("%032x", uint64(now))
	spanID := fmt.Sprintf("%016x", uint64(now))
	traceBody := fmt.Sprintf("{\"resourceSpans\":[{\"resource\":{\"attributes\":[{\"key\":\"service.name\",\"value\":{\"stringValue\":\"%s\"}}]},\"scopeSpans\":[{\"spans\":[{\"traceId\":\"%s\",\"spanId\":\"%s\",\"name\":\"%s\",\"kind\":1,\"startTimeUnixNano\":\"%d\",\"endTimeUnixNano\":\"%d\",\"status\":{\"code\":1}}]}]}]}", t.service, traceID, spanID, name, now, now+1e6)
	logBody := fmt.Sprintf("{\"resourceLogs\":[{\"resource\":{\"attributes\":[{\"key\":\"service.name\",\"value\":{\"stringValue\":\"%s\"}}]},\"scopeLogs\":[{\"logRecords\":[{\"timeUnixNano\":\"%d\",\"severityNumber\":9,\"body\":{\"stringValue\":\"request\"}}]}]}]}", t.service, now)
	if err := t.post("/v1/traces", traceBody); err != nil {
		t.status = "degraded"
	}
	_ = t.post("/v1/logs", logBody)
}

func (t *telemetry) increment(name string) {
	now := time.Now().UnixNano()
	body := fmt.Sprintf("{\"resourceMetrics\":[{\"resource\":{\"attributes\":[{\"key\":\"service.name\",\"value\":{\"stringValue\":\"%s\"}}]},\"scopeMetrics\":[{\"metrics\":[{\"name\":\"%s\",\"sum\":{\"aggregationTemporality\":2,\"isMonotonic\":true,\"dataPoints\":[{\"asInt\":\"1\",\"startTimeUnixNano\":\"%d\",\"timeUnixNano\":\"%d\"}]}}]}]}]}", t.service, name, now, now)
	if err := t.post("/v1/metrics", body); err != nil {
		t.status = "degraded"
	}
}

func (t *telemetry) record(name string, value float64) {
	now := time.Now().UnixNano()
	body := fmt.Sprintf("{\"resourceMetrics\":[{\"resource\":{\"attributes\":[{\"key\":\"service.name\",\"value\":{\"stringValue\":\"%s\"}}]},\"scopeMetrics\":[{\"metrics\":[{\"name\":\"%s\",\"histogram\":{\"aggregationTemporality\":2,\"dataPoints\":[{\"count\":\"1\",\"sum\":%g,\"timeUnixNano\":\"%d\"}]}}]}]}]}", t.service, name, value, now)
	if err := t.post("/v1/metrics", body); err != nil {
		t.status = "degraded"
	}
}

func (t *telemetry) post(path, body string) error {
	req, err := http.NewRequestWithContext(context.Background(), http.MethodPost, t.base+path, strings.NewReader(body))
	if err != nil {
		return err
	}
	req.Header.Set("content-type", "application/json")
	res, err := t.client.Do(req)
	if err != nil {
		return err
	}
	defer res.Body.Close()
	_, _ = io.Copy(io.Discard, res.Body)
	return nil
}
